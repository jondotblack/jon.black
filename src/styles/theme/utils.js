import { baseFontSize } from './constants';

/**
 * Internal helper function
 * Convert pixel value to relative unit
 * @param {string} unit
 * @param {number} base
 * @returns {function(number): string}
 */
export function pxToRelUnit(unit, base) {
  return (x) => `${parseFloat(x) / base}${unit}`;
}

// Relative unit functions
export const pxToEm = pxToRelUnit('em', baseFontSize);
export const pxToRem = pxToRelUnit('rem', baseFontSize);
export const mapToEm = (arr) => arr.map(pxToEm);
export const mapToRem = (arr) => arr.map(pxToRem);

/**
 * Maps breakpoints dictionary, and creates utilities for
 * constructing media queries with those keys.
 *
 * This function takes an object structured
 * like so:
 * ```
 * {
 *   sm: 640px,
 *   md: 960px,
 *   lg: 1280px
 * }
 *
 * The mapping function will be applied to each of the values, and the keys
 * will be used to generate the correct media query string.
 *
 * For example, `media.lessThan('sm')` will produce a string like
 * `(max-width: 640px)` that can be used for styled components.
 *
 * @param {object} breakpoints - the breakpoints object
 */
export function mapMedia(breakpoints) {
  const mapQuery = (func, keys, ...args) => {
    const values = args.flat().map(func);
    const queries = keys.map((key, i) => `(${key}: ${values[i]})`);
    return queries.join(' and ');
  };

  const breakpoint = (key) => breakpoints[key];
  const [min, max] = ['min-width', 'max-width'];

  return {
    // Returns breakpoint value
    breakpoint,
    // Returns true if breakpoint is less than key, i.e. lessThan('sm')
    lessThan: (key) => mapQuery(breakpoint, [max], key),
    // Returns true if breakpoint is more than key, i.e. moreThan('sm')
    moreThan: (key) => mapQuery(breakpoint, [min], key),
    // Returns true if breakpoint is between keys, i.e. between('sm', 'lg')
    between: (...keys) => mapQuery(breakpoint, [min, max], ...keys),
  };
}
