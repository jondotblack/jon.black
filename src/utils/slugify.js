import deburr from 'lodash/deburr';

/* eslint-disable no-useless-escape */
export const slugify = (string) =>
  deburr(string)
    .trim()
    .toLowerCase()
    .replace(/\s+/g, '-') // replace spaces with -
    .replace(/&/g, 'and') // replace & with 'and'
    .replace(/\./g, '-') // replace . with -
    .replace(/([^a-zA-Z0-9\._-]+)/, '');
/* eslint-enable */
