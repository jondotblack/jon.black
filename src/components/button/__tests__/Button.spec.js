import React from 'react';

import renderer from 'react-test-renderer';
import { fireEvent, render, screen } from 'utils/testing';

import { Button } from '../index';

describe('Button', () => {
  const mockOnClick = jest.fn();
  const text = 'call to action';

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render as a button correctly with all props', () => {
    const tree = renderer.create(<Button>{text}</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render as a link correctly with all props', () => {
    const tree = renderer.create(<Button href="https://www.helloroot.co">{text}</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fire onClick event correctly', () => {
    render(<Button onClick={mockOnClick}>{text}</Button>);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('should not fire onClick event when disabled', () => {
    render(
      <Button onClick={mockOnClick} disabled>
        {text}
      </Button>,
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(button).toHaveTextContent(text);
    expect(mockOnClick).toHaveBeenCalledTimes(0);
  });

  it('should not fire onClick event or display children when loading', () => {
    render(
      <Button onClick={mockOnClick} loading>
        {text}
      </Button>,
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(button).not.toHaveTextContent(text);
    expect(mockOnClick).toHaveBeenCalledTimes(0);
  });
});
