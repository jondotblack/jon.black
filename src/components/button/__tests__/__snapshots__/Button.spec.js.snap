// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render as a button correctly with all props 1`] = `
.emotion-0 {
  display: inline-block;
  padding: 0.375rem 0.75rem;
  font-size: 18px;
  line-height: 1.5;
  color: #40413b;
  text-transform: uppercase;
  text-align: center;
  vertical-align: middle;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: color 0.15s ease-in-out,background-color 0.15s ease-in-out,border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  transition: color 0.15s ease-in-out,background-color 0.15s ease-in-out,border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
}

.emotion-0:focus {
  outline: none;
}

<button
  className="emotion-0 emotion-1"
  disabled={false}
  onClick={[Function]}
>
  call to action
</button>
`;

exports[`Button should render as a link correctly with all props 1`] = `
.emotion-0 {
  display: inline-block;
  padding: 0.375rem 0.75rem;
  font-size: 18px;
  line-height: 1.5;
  color: #40413b;
  text-transform: uppercase;
  text-align: center;
  vertical-align: middle;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 0.25rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: color 0.15s ease-in-out,background-color 0.15s ease-in-out,border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
  transition: color 0.15s ease-in-out,background-color 0.15s ease-in-out,border-color 0.15s ease-in-out,box-shadow 0.15s ease-in-out;
}

.emotion-0:focus {
  outline: none;
}

<a
  className="emotion-0 emotion-1"
  disabled={false}
  href="https://www.helloroot.co"
  onClick={[Function]}
>
  call to action
</a>
`;
